assembly/src/assembler/mod.rs

```diff
@@ -62,18 +60,31 @@ pub struct Assembler {
 // ------------------------------------------------------------------------------------------------
 /// Constructors
 impl Assembler {
-    /// Start building an [Assembler]
-    pub fn new() -> Self {
-        Self::default()
-    }
+    /// Start building an [`Assembler`] with a kernel defined by the provided [CompiledLibrary].
+    ///
+    /// # Errors
+    /// Returns an error if:
+    /// - The library does not contain a kernel module.
+    /// - The library contains non-kernel modules.
```

Comment:
While this is necessary when we use panics, I think we should avoid doing this when returning errors. Ideally, the error type encodes all the possible errors from a given function, and no other. And we can use thiserror's `#[from]` annotation to easily define error trees that work seemlessly with the `?` operator.

So comments like this duplicate the information provided by the error type, are tedious to maintain, and risk falling out of sync with the code - hence I would suggest avoiding them.

Reply:
So I want to push back on that in this case @plafer, for the simple reason that I don't think we should be returning anything other than `Report` from the `Assembler`. The assembler is a bit different than other areas of the Miden codebase. We generally _don't_ want concrete error types here, and instead should use `Report` (essentially a boxed `Diagnostic` trait object that implements `std::error::Error`).

This is because errors produced by the assembler are meant to be rendered to stdout (or printed to a string), so that they can include all of the extra information contained in the gathered diagnostic(s) (such as source spans when relevant, codes/URIs, etc.). While there may be cases where we have "simple" errors that don't leverage those capabilities (such as this function), those are generally the exception, rather than the rule, and there is little benefit to exposing those internal details to consumers of the `Assembler` in the form of a one-off error type. For things like this, internal error types can be propagated using `Report::wrap_err` (either using the method, or via the `WrapErr` trait, exported from `miden_assembly::diagnostics`), simple diagnostics can be created from raw `Error` impls or `Display` impls using `Report::new` or `Report::msg` respectively, and there are macros available to construct full diagnostics ad-hoc as instances of `Report`.

So while I agree that we don't want to document redundant information in the doc comments, in this case I think it is appropriate to document what type of issues can cause an error to be reported due to user error, since all the user will get is a `Report`.

Reply:
I think I see @bitwalker's point. Since all errors generated by the assembler (including failure to instantiate the assembler) are meant to be read by humans, as opposed to errors being handled programmatically, then maintaining error enums is tedious for little benefit.

This is the same reasoning behind the `anyhow` library (and after looking more into miette, it seems like they are intrically linked). So actually I agree that we should use `Result<_, Report>` everywhere (which I now realize is analogous to `anyhow::Result`).

> Regarding having an error section which describes error conditions: I think it is pretty useful and also Rust standard library does this sometimes as well

Interesting, didn't realize that before. But also if we are indeed switching to using `Report` everywhere, then it definitely makes sense to keep them.

Reply:
After introducing `KernelLibrary` struct, the need for `AssemblerError` went away. So, I removed it.